apps/total-typescript/src/hooks/use-progress.ts

```diff
@@ -99,206 +69,65 @@ export const getNextSectionForProgress = ({
-  exercises: Exercise[]
-  nextExercise?: Lesson | null
-}) => {
-  const isCompleted = completedResources.length === resources.length
-  const percentCompleted = Math.round(
-    (completedResources.length / resources.length) * 100,
-  )
-
-  const hasProgress = !isEmpty(
-    find(completedResources, {
-      // if explainer
-      lessonSlug: get(find(exercises, {slug: nextExercise?.slug}), 'slug'),
-    }) ||
-      find(completedResources, {
-        lessonSlug: get(
-          find(exercises, {slug: nextExercise?.slug}),
-          'solution.slug',
-        ),
-      }),
-  )
-
-  const lastCompletedSolution = getLastCompletedSolution({
-    resources,
-    completedResources,
-  })
-
-  return {isCompleted, percentCompleted, lastCompletedSolution, hasProgress}
-}
-
 export const useModuleProgress = ({module}: {module: Module}) => {
   const {sections} = module
   const {data: userProgress = [], status: userProgressStatus} =
     trpcSkillLessons.progress.get.useQuery()
 
-  const {exercises, resources} = extractExercisesAndResource(sections)
-
-  // storing this in state allows to refetch nextExercise
-  // in case the original next is already completed
-  const [lastCompletedExercise, setLastCompletedExercise] = React.useState<
-    Lesson | null | undefined
-  >(first(exercises))
-
-  const completedResources = getCompletedResources({userProgress, resources})
+  const lessons = getAllLessons(sections)
+  const {completedLessons, lastCompletedLesson} = getCompletedLessons({
+    userProgress,
+    lessons,
+  })
 
   const nextSection = getNextSectionForProgress({
-    userProgress,
-    resources,
-    lastCompletedExercise,
+    lastCompletedLesson,
     sections,
     module,
   })
 
-  const {data: nextExercise, status: nextExerciseStatus} =
+  const {data: nextLesson, status: nextLessonStatus} =
     trpcSkillLessons.lessons.getNextLesson.useQuery({
       type: 'solution',
-      slug: lastCompletedExercise?.slug,
+      slug: lastCompletedLesson?.slug,
       section: nextSection?.slug,
       module: module.slug.current,
     })
 
-  const {isCompleted, percentCompleted, lastCompletedSolution, hasProgress} =
-    getProgress({
-      resources,
-      completedResources,
-      exercises,
-      nextExercise,
-    })
-
-  React.useEffect(() => {
-    const lessons = nextSection?.lessons
-    if (hasProgress && nextSection) {
-      setLastCompletedExercise(
-        lessons
-          ? lessons[indexOf(lessons, find(lessons, {slug: nextExercise?.slug}))]
-          : null,
-      )
-    }
-    if (userProgressStatus === 'success' && lastCompletedSolution) {
-      setLastCompletedExercise(
-        find(exercises, {
-          slug: lastCompletedSolution.slug,
-        }) ||
-          find(exercises, {
-            solution: lastCompletedSolution,
-          }),
-      )
-    }
-  }, [
-    exercises,
-    hasProgress,
-    lastCompletedSolution,
-    nextExercise?.slug,
-    nextSection,
-    userProgressStatus,
-  ])
```

Comment:
once I started removing stuff this felt like it was not needed, but it may still be just wanted to reset a bit and rethink

Reply:
I was thinking that a better approach would be to have `getNextLessonForProgress` trpc query which would always give us a correct "next up" lesson in correlation with user progress by skipping already completed ones in sequence.